core/src/spclient.rs

```diff
@@ -1 +1,255 @@
+                        }
+                    }
+                    _ => break, // if we can't build the request now, then we won't ever
+                }
+            }
+
+            // When retrying, avoid hammering the Spotify infrastructure by sleeping a while.
+            // The backoff time is chosen randomly from an ever-increasing range.
+            let max_seconds = u64::pow(tries as u64, 2) * 3;
+            let backoff = Duration::from_secs(rand::thread_rng().gen_range(1..=max_seconds));
+            warn!(
+                "Unable to complete API request, waiting {} seconds before retrying...",
+                backoff.as_secs(),
+            );
+            debug!("Error was: {:?}", last_response);
+            tokio::time::sleep(backoff).await;
+        }
+
+        last_response
+    }
+
+    pub async fn put_connect_state(
+        &self,
+        connection_id: String,
+        state: protocol::connect::PutStateRequest,
+    ) -> SpClientResult {
+        let endpoint = format!("/connect-state/v1/devices/{}", self.session().device_id());
+
+        let mut headers = HeaderMap::new();
+        headers.insert("X-Spotify-Connection-Id", connection_id.parse()?);
+
+        self.protobuf_request("PUT", &endpoint, Some(headers), &state)
+            .await
+    }
+
+    pub async fn get_metadata(&self, scope: &str, id: SpotifyId) -> SpClientResult {
+        let endpoint = format!("/metadata/4/{}/{}", scope, id.to_base16());
+        self.request("GET", &endpoint, None, None).await
+    }
+
+    pub async fn get_track_metadata(&self, track_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("track", track_id).await
+    }
+
+    pub async fn get_episode_metadata(&self, episode_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("episode", episode_id).await
+    }
+
+    pub async fn get_album_metadata(&self, album_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("album", album_id).await
+    }
+
+    pub async fn get_artist_metadata(&self, artist_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("artist", artist_id).await
+    }
+
+    pub async fn get_show_metadata(&self, show_id: SpotifyId) -> SpClientResult {
+        self.get_metadata("show", show_id).await
+    }
+
+    // TODO: Not working at the moment, always returns 400.
+    pub async fn get_lyrics(&self, track_id: SpotifyId) -> SpClientResult {
+        // /color-lyrics/v2/track/22L7bfCiAkJo5xGSQgmiIO/image/spotify:image:ab67616d0000b273d9194aa18fa4c9362b47464f?clientLanguage=en
+        // https://spclient.wg.spotify.com/color-lyrics/v2/track/{track_id}/image/spotify:image:{image_id}?clientLanguage=en
+        let endpoint = format!("/color-lyrics/v2/track/{}", track_id.to_base16());
+
+        let mut headers = HeaderMap::new();
+        headers.insert("Content-Type", "application/json".parse()?);
+
+        self.request("GET", &endpoint, Some(headers), None).await
```

Comment:
As mentioned on the Gitter chat, this isn't working at the moment. Not even when using that full example URL on line 227. It returns a HTTP status code 400 every time. Help is welcome.

 * `POST` doesn't work and returns 405 (I tried posting with an `{}` empty JSON request).
 * Removing `Content-Type` doesn't work either (without it, it should return a binary response).

Should it be requested from a different access point? Or with access token other than `playlist-read`? I don't know.

Reply:
Fixed, it was version-gated and required an official `User-Agent`.